util/src/test/java/io/grpc/util/AdvancedTlsX509TrustManagerTest.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.internal.testing.TestUtils;
+import io.grpc.testing.TlsTesting;
+import java.io.File;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Handler;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link AdvancedTlsX509TrustManager}. */
+@RunWith(JUnit4.class)
+public class AdvancedTlsX509TrustManagerTest {
+
+  public static final String CA_PEM_FILE = "ca.pem";
+  public static final String SERVER_0_PEM_FILE = "server0.pem";
+  private File caCertFile;
+  private File serverCert0File;
+
+  private X509Certificate[] caCert;
+  private X509Certificate[] serverCert0;
+
+  private ScheduledExecutorService executor;
+
+  @Before
+  public void setUp() throws IOException, GeneralSecurityException {
+    executor = Executors.newSingleThreadScheduledExecutor();
+    caCertFile = TestUtils.loadCert(CA_PEM_FILE);
+    caCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(CA_PEM_FILE));
+    serverCert0File = TestUtils.loadCert(SERVER_0_PEM_FILE);
+    serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+  }
+
+  @After
+  public void tearDown() {
+    MoreExecutors.shutdownAndAwaitTermination(executor, 5, TimeUnit.SECONDS);
+  }
+
+  @Test
+  public void credentialSetting() throws Exception {
```

Comment:
Same style nits here RE test structure as in the key manager PR, but we can defer to Eric if you prefer.

Reply:
The name here is pretty weak. `updateTrustCredentials_replacesIssuers` would seem better. Although it seems `updateTrustCredentialsFromFile(File)` should also be run again at the end if the point is to notice the replacement. Otherwise, I'd have expected three separate instances of AdvancedTlsX509TrustManager to be used (and three separate tests then seems fair).